src/Microsoft.ManagedZlib/src/OutputWindow.cs

```diff
@@ -0,0 +1,198 @@
+                Array.Copy(_window, copyStart, _window, _lastIndex, length);
+                _lastIndex += length;
+            }
+            else
+            {
+                // The referenced string may overlap the current
+                // position; for example, if the last 2 bytes decoded have values
+                // X and Y, a string reference with <length = 5, distance = 2>
+                // adds X,Y,X,Y,X to the output stream.
+                while (length-- > 0)
+                {
+                    _window[_lastIndex++] = _window[copyStart++];
+                }
+            }
+        }
+        else
+        {
+            // copy byte by byte
+            while (length-- > 0)
+            {
+                _window[_lastIndex++] = _window[copyStart++];
+                _lastIndex &= WindowMask;
+                copyStart &= WindowMask;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Copy up to length of bytes from input directly.
+    /// This is used for uncompressed block, after passing through Decode().
+    /// </summary>
+    public int CopyFrom(InputBuffer input, int length)
+    {
+        // Remember: We are copying the data to the second portion of the window -look-ahead-
+        // So bytesUsed might be the size of the first portion 32K or 64K, at this point.
+        //AvailableBytes is usually the size of the underlying buffer of the stream (for DeflateStream)
+        //This in particular is for copying the input respecting the I/O boundaries. 
+        // It will lead us to either copy LEN bytes or just the amount available in the output window
+        // (taking into account the byte boundaries)
+        /// <summary> 
+        /// Either how much input is available or how much free space in the output buffer we have. 
+        /// length = Amount decompressed that we are trying to put in the output window
+        /// </summary>
+        length = Math.Min(Math.Min(length, WindowSize - _bytesUsed), input.AvailableBytes);
+        int copied;
+
+        // We might need wrap around to copy all bytes.
+        int spaceLeft = WindowSize - _lastIndex;
+        if (length > spaceLeft) //Checking is in the boundaries
+        {
+            // copy the first part
+            copied = input.CopyTo(_window, _lastIndex, spaceLeft);
+            if (copied == spaceLeft)
+            {
+                // only try to copy the second part if we have enough bytes in input
+                copied += input.CopyTo(_window, 0, length - spaceLeft);
+            }
+        }
+        else
+        {
+            // only one copy is needed if there is no wrap around.
+            copied = input.CopyTo(_window, _lastIndex, length);
+        }
+
+        _lastIndex = (_lastIndex + copied) & WindowMask; //To keep it withting the window sizw boundary
+        _bytesUsed += copied;
+        return copied; // Me imagino que esto es lo que espera el defalte stream como "bytesRead"
```

Comment:
Please translate the comments to english so you can get a proper review from all folks.